File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/ContextCompressor.fs
namespace Skight.AgentPlatform.FSharp

module ContextCompressor =

    /// Estimates token budget usage and compresses context when exceeding threshold (default 80%)
    let compress (thresholdRatio: float) (limit: int) (msgs: AgentMessage list) : AgentMessage list =
        let triggerThreshold = int (float limit * thresholdRatio)
        if msgs.Length <= triggerThreshold then
            msgs
        else
            printfn "  [Context Compaction Engine] History size (%d) exceeds threshold (%d of limit %d). Compacting..." msgs.Length triggerThreshold limit
            let systemPrompt = msgs.Head
            let keepRecentCount = System.Math.Max(3, limit / 3)
            let recentMessages =
                msgs
                |> List.skip (msgs.Length - keepRecentCount)
                |> List.skipWhile (function | ToolMessage _ -> true | _ -> false)

            let trimmedCount = msgs.Length - recentMessages.Length - 1
            let summaryContent =
                sprintf "[TURN SUMMARY]: %d past conversation turns were compacted to maintain token budget. Key focus is retained in recent context." trimmedCount

            let summaryMsg = SystemMessage summaryContent
            systemPrompt :: summaryMsg :: recentMessages

File: implementations/csharp/src/Skight.AgentPlatform/ContextCompressor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Azure.AI.OpenAI;

namespace Skight.AgentPlatform
{
    public static class ContextCompressor
    {
        public static List<ChatRequestMessage> Compress(double thresholdRatio, int limit, List<ChatRequestMessage> messages)
        {
            if (messages == null || messages.Count == 0) return new List<ChatRequestMessage>();

            int triggerThreshold = (int)(limit * thresholdRatio);
            if (messages.Count <= triggerThreshold)
            {
                return messages;
            }

            Console.WriteLine($"  [Context Compaction Engine] History size ({messages.Count}) exceeds threshold ({triggerThreshold} of limit {limit}). Compacting...");

            var systemPrompt = messages[0];
            int keepRecentCount = Math.Max(3, limit / 3);

            var recentMessages = messages
                .Skip(Math.Max(0, messages.Count - keepRecentCount))
                .SkipWhile(m => m is ChatRequestToolMessage)
                .ToList();

            int trimmedCount = messages.Count - recentMessages.Count - 1;
            string summaryContent = $"[TURN SUMMARY]: {trimmedCount} past conversation turns were compacted to maintain token budget. Key focus is retained in recent context.";

            var summaryMsg = new ChatRequestSystemMessage(summaryContent);

            var result = new List<ChatRequestMessage> { systemPrompt, summaryMsg };
            result.AddRange(recentMessages);
            return result;
        }
    }
}
